src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,164 @@
+pragma solidity =0.8.12;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../interfaces/IPaymentCoordinator.sol";
+import "../libraries/Merkle.sol";
+import "@openzeppelin-upgrades/contracts/proxy/utils/Initializable.sol";
+import "@openzeppelin-upgrades/contracts/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "forge-std/Test.sol";
+
+
+
+/**
+ * @title Contract used to coordinate payments from AVSs to operators and in particular the subsequency splitting of earnings from operators to stakers
+ * @author Layr Labs, Inc.
+ */
+contract PaymentCoordinator is 
+    IPaymentCoordinator,
+    Initializable,
+    OwnableUpgradeable,
+    Test
+{
+    using SafeERC20 for IERC20;
+
+    /// @notice address approved to post new Merkle roots
+    address public rootPublisher;
+
+    uint256 public merkleRootActivationDelay;
+
+    /// @notice maximum BIPS
+    uint256 public constant MAX_BIPS = 10000;
+
+    /// @notice Array of roots of posted Merkle trees, as well as associated data like tree height
+    MerkleRootPost[] public merkleRoots;
+
+    /// @notice Mapping token => recipient => cumulative amount *claimed*
+    mapping(IERC20 => mapping(address => uint256)) public cumulativeTokenAmountClaimedByRecipient;
+
+    /// @notice Mapping token => cumulative amount earned by EigenLayer
+    mapping(IERC20 => uint256) public cumulativeEigenLayerTokeEarnings;
+
+     /// @notice Constant that defines the share EigenLayer takes of all payments, in basis points
+     uint256 public eigenLayerShareBIPs;
+
+    // TODO: better define this event?
+    event PaymentReceived(address indexed receivedFrom, Payment payment);
+
+    // @notice Emitted when a new Merkle root is posted
+    event NewMerkleRootPosted(MerkleRootPost merkleRootPost);
+
+    event PaymentClaimed(MerkleLeaf merkleLeaf);
+
+    modifier onlyRootPublisher {
+        require(msg.sender == rootPublisher, "PaymentCoordinator: Only rootPublisher");
+        _;
+    }
+
+    function initialize(address _initialOwner, address _rootPublisher, uint256 _eigenlayerShareBips, uint256 _merkleRootActivationDelay)
+        external
+        initializer
+    {
+        _transferOwnership(_initialOwner);
+        _setRootPublisher(_rootPublisher);
+        _setMerkleRootActivationDelay(_merkleRootActivationDelay);
+        _setEigenLayerShareBIPS(_eigenlayerShareBips);
+    }
+
+
+    /**
+     * @notice Makes a payment of sum(amounts) paid in `token`, for `operator`'s contributions to an AVS,
+     * between `startBlockNumber` (inclusive) and `endBlockNumber` (inclusive)
+     * @dev Transfers the total payment from the `msg.sender` to this contract, so the caller must have previously approved
+     * this contract to transfer at least sum(`amounts`) of `token`
+     * @notice Emits a `PaymentReceived` event
+     */
+    function makePayment(Payment calldata payment) external{
+        uint256 sumAmounts;
```

Comment:
I think this function should also check that the lengths of `amounts` and `quorums` match. seems important to enforce this, and I think this is the proper place?

Reply:
Yeah.  Do we need quorums in the first place as far as the eigenlayer payments is concerned?  Seems like no?  Unless we do need it to be emitted for some offchain stuff

Reply:
yes, this is all formatted with off-chain calculations in mind